openjdk/barriers/mmtkSATBBarrier.cpp

```diff
@@ -0,0 +1,533 @@
+// Workaround the fact that LIR_Assembler::as_Address is private.
+#define private public
+
+#include "precompiled.hpp"
+#include "mmtkSATBBarrier.hpp"
+#include "runtime/interfaceSupport.inline.hpp"
+
+#define SOFT_REFERENCE_LOAD_BARRIER true
+
+constexpr int kUnloggedValue = 1;
+
+static inline intptr_t side_metadata_base_address() {
+  return SATB_METADATA_BASE_ADDRESS;
+}
+
+void MMTkSATBBarrierSetRuntime::load_reference(DecoratorSet decorators, oop value) const {
+#if SOFT_REFERENCE_LOAD_BARRIER
+  if (CONCURRENT_MARKING_ACTIVE == 1 && value != NULL)
+    ::mmtk_load_reference((MMTk_Mutator) &Thread::current()->third_party_heap_mutator, (void*) value);
+#endif
+};
+
+void MMTkSATBBarrierSetRuntime::object_probable_write(oop new_obj) const {
+  // We intentionally leave this method blank.
```

Comment:
Wait is this actually sound? I thought this can be called for de-optimizations as well?